### PR TITLE
feat: Implement pseudo_random.cairo helper for deterministic randomness

### LINKED ISSUE #79: 📝 [FEAT]: Implement pseudo_random.cairo helper for deterministic randomness
### 🔍 Scope
Add a pseudo-random number generator helper in the helpers folder to provide deterministic randomness for Dojo games, allowing developers to generate random values for game mechanics.

### 🔄 How to Implement the Issue
- Create a new file `pseudo_random.cairo` in the helpers folder
- Implement the `PseudoRandom` module with the `generate_random_u8` function
- Use Starknet block data (timestamp, block number) as entropy sources
- Implement test cases to verify functionality
- Document usage examples for game developers

### ✅ Acceptance Criteria
- File is correctly placed in helpers folder with proper naming
- PseudoRandom trait is implemented with generate_random_u8 function
- Function generates values within specified min/max range
- Function produces consistent outputs for same inputs
- Comprehensive tests verify functionality and edge cases
- Code is properly documented with docstrings
- Function uses block data for pseudo-randomness

### 📸 Additional Context
Deterministic randomness is essential for onchain games. This helper will enable developers to generate random stats, outcomes, and gameplay elements in a predictable yet varied way. The implementation should follow the provided code structure and use Pedersen hash for generating random values based on block data.

### Contribution Guidelines
- Follow the contribution guidelines in the project's README
- Ensure gas efficiency in the implementation
- Add thorough documentation explaining limitations of onchain randomness

### PR DESCRIPTION
**Pull request by truthixify**
# Pull Request Overview

## 📝 Summary
This PR implements a pseudo-random number generator for generating random u8 values within a specified range, leveraging block data for entropy. It ensures that the generated random values are deterministic for the same block and fit within the given min and max range. Additionally, tests for various use cases, including edge cases, have been added.

### Related Issues
- Closes #79 .

### Type of Change
Mark with an `x` all the checkboxes that apply (like `[x]`).
- [ ] 📝 Documentation (updates to README, docs, or comments)
- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 👌 Enhancement (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)

### What's Changed
- Implemented a generate_random_u8 function that generates a pseudo-random u8 based on block data (timestamp and block number).
- Added assertions to ensure the generated number falls within the desired range.
- Enhanced test coverage by introducing edge cases and validation tests.

### Implementation Details
- The function uses the Pedersen hash function to create entropy from the Starknet block timestamp and block number.
- It converts the generated hash into a u256, then reduces it to a u8 that fits within the specified range (min, max).
- Unit tests were added to check for expected behavior and edge cases like range boundaries and invalid inputs.

### Technical Notes
- Ensure that the Pedersen trait is imported correctly and configured in the project.

## 🔧 Tests Results
Describe the tests you performed to verify your changes.

### Test Coverage
- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Testing

### Evidence
Provide relevant evidence of testing (screenshots, test outputs, etc.).

image

### Testing Notes
- Tests were written to validate that random numbers stay within the specified range.
- Edge cases such as equal min and max were handled with a panic check.

## 🔜 Next Steps
- Review performance for scalability if random number generation is used in high-frequency scenarios.
- Add further integration tests to check for real-world entropy quality or uniqueness of random values across different blocks.
- Consider adding an optional seed parameter for more control over the randomness if needed.

### CODE REVIEW

File: backend/dojo_examples/combat_game/src/helpers/pseudo_random.cairo [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    /// * `max`: The maximum value (inclusive) of the random number range.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if `min >= max`.
+    ///
+    /// # Returns
+    ///
+    /// A pseudo-random `u8` value within the specified range.
+    pub fn generate_random_u8(min: u8, max: u8) -> u8 {
```
Comment (coxmars) [inline backend/dojo_examples/combat_game/src/helpers/pseudo_random.cairo:26]: Hi @truthixify while your implementation works for basic cases, it lacks flexibility needed for our use cases:

1. Current implementation produces the same number for all calls within a block.
2. Need to add unique_id and salt parameters to generate different values in same block.

Suggested implementation:

```cairo
pub fn generate_random_u8(
    unique_id: u16, 
    salt: u16,
    min: u8, 
    max: u8
) -> u8 {
    assert!(min < max, "min must be less than max");
    
    // Get entropy values
    let block_timestamp = get_block_timestamp();
    let block_number = get_block_number();
    
    // Create unique seeds with your parameters
    let timestamp_seed: felt252 = block_timestamp.into() + salt.into();
    let block_seed: felt252 = block_number.into() + unique_id.into();
    let combined_seed: felt252 = timestamp_seed + (unique_id * salt).into();
    
    // Hash for randomness
    let hash_input = combined_seed + block_seed;
    let hash_state = PedersenTrait::new(0);
    let hash_state = hash_state.update(hash_input);
    let hash = hash_state.finalize();
    
    // Convert to range
    let range: u64 = (max - min + 1).into();
    let hash_u256: u256 = hash.into();
    let mod_value: u64 = (hash_u256 % range.into()).try_into().unwrap();
    let random_value: u8 = mod_value.try_into().unwrap() + min;
    
    assert!(random_value >= min && random_value <= max, "Random value out of range");
    return random_value;
}
```

And adjust the uni tests with these changes please and we can merge.

**Review by coxmars — CHANGES_REQUESTED** [review]
Apply the changes

File: backend/dojo_examples/combat_game/src/helpers/pseudo_random.cairo [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        let min: u8 = 0;
+        let max: u8 = 255;
+        let unique_id: u16 = 12345;
+        let salt: u16 = 6789;
+        let result = generate_random_u8(unique_id, salt, min, max);
+
+        // Assert the result is within the full `u8` range
+        assert!(result >= min && result <= max, "Value should be in [0, 255]");
+    }
+}
```
Comment (coxmars) [inline backend/dojo_examples/combat_game/src/helpers/pseudo_random.cairo:144]: Add a new line at the end of this file please

**Review by coxmars — CHANGES_REQUESTED** [review]
@truthixify apply that minor fix and resolve your conflicts and we can merge

**Review by jimenezz22 — APPROVED** [review]
